File: prob2.py
# Time Complexity :  O(1)
# Space Complexity :O(n)
# Passed on Leetcode: yes

import queue

class TrieNode():
    def __init__(self):
        self.flag = False
        self.arr = [None for _ in range(26)]

class Trie():
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, word):
        node = self.root
        for char in word:
            idx = ord(char) - ord('a')
            if node.arr[idx] == None:
                node.arr[idx] = TrieNode()
            node = node.arr[idx]
        node.flag = True

class Solution:
    def longestWord(self, words: List[str]) -> str:
        trie = Trie()
        for val in words:
            trie.insert(val)
        
        q = queue.Queue()
        s = queue.Queue()
        q.put(trie.root)
        s.put('')
        n = ''
        while not q.empty():
            node = q.get()
            st = s.get()
            for idx in range(25, -1, -1):
                child = node.arr[idx]
                if child != None and child.flag == True:
                    q.put(child)
                    n = st + chr(ord('a') + idx)
                    s.put(n)

        return n

File: prob2.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes



class Solution {
    public String longestWord(String[] words) {
        Set<String> set = new HashSet<>();

        for(int i = 0; i < words.length; i++){
            set.add(words[i]);
        }

        Queue<String> q = new LinkedList<>();

        q.add("");
        String prev = "";

        while(!q.isEmpty()){
            String s = q.poll();
            prev = s;
            for(int i = 25; i >=0 ; i--){
                String key = s + String.valueOf((char) ('a' + i));
                if(set.contains(key)){
                    q.add(key);;
                }
            }
        }
        return prev;
    }
}